sycl/source/detail/event_impl.cpp

```diff
@@ -100,6 +102,9 @@ void event_impl::wait(
     waitInternal();
   else if (MCommand)
     detail::Scheduler::getInstance().waitForEvent(std::move(Self));
+  if (MCommand && !SYCLConfig<SYCL_DISABLE_EXECUTION_GRAPH_CLEANUP>::get())
```

Comment:
What would I need `SYCL_DISABLE_EXECUTION_GRAPH_CLEANUP` for?

Reply:
It's intended for debugging SYCL runtime, alongside the `SYCL_PRINT_EXECUTION_GRAPH` environment variable. Viewing the whole graph rather than only the nodes left after cleanup might be useful.

Reply:
So, what is the problem with dumping the graph before and after clean-up?
Do I understand correctly that enabling this variables leads to memory leaks in scheduler?

Reply:
> So, what is the problem with dumping the graph before and after clean-up?

Dumping the graph before and after would not let you view the whole execution graph if the cleanup is performed multiple times.

> Do I understand correctly that enabling this variables leads to memory leaks in scheduler?

Not quite. The issue this patch resolves is not exactly memory leaks, since even now all commands are correctly deallocated during memory object destruction. The problem is that if the user submits a large enough number of command groups using one memory object, the SYCL runtime consumes way too much memory. So, to resolve this problem, rather than waiting for the memory object destruction, we release already finished commands earlier.

Enabling the environment variable brings back this problem, but I would imagine that printing the whole execution graph would not be very informative in such problematic cases anyway.